Use baselayerchange/overlayadd/overlayremove instead of layeradd/layerremove

deevroman opened the pull request:
### Description
This is fix #5466 After merging https://github.com/openstreetmap/leaflet-osm/pull/43 flame graph looks like this:

<img width="1208" src="https://github.com/user-attachments/assets/435540a1-ebe4-40e6-8051-2d881e3ed28e" />

https://share.firefox.dev/4a2Wc9Z

Most of the time is spent on two handlers that update information about the current location and elements in the right sidebar. 

Unfortunately, replacing `layeradd layerremove` with `baselayerchange overlaylayerchange` (https://github.com/openstreetmap/openstreetmap-website/issues/5466#issuecomment-2571570472) is difficult, since in addition to the Map Data layer, other elements may be displayed on the map. For example, the currently selected object on the map. It is not clear how to separate changes to Map Data objects and changes to the selected element. It is necessary to track changes in the active element, otherwise at least the Edit button will break.

~~Therefore, I implemented a little ugly, but safer in terms of breaking business logic. When adding the Map Data layer, I set the pause flag in the slowdown handlers. While the Map Data is rendering, the user cannot interact with the interface, so it seems safe to pause the event handlers.~~

Total 15s vs 0.5s when rendering 5000 elements.

<img width="1207" src="https://github.com/user-attachments/assets/13a099d0-0918-4c32-840d-24faf5a1cb8c" />

https://share.firefox.dev/4h4IK7C

AntonKhorev commented:
> For example, the currently selected object on the map. It is not clear how to separate changes to Map Data objects and changes to the selected element. It is necessary to track changes in the active element, otherwise at least the Edit button will break.

`addObject`/`removeObject` in `leaflet.map.js` could fire some event for active object change. The listener that calls `updateLinks` could listen to that event.

deevroman commented:
I also saw a problem in PR: when you click the Map Data checkbox, the cookie with the current location is no longer updated

deevroman commented:
Okay, firing events from `addObject` / `removeObject` turned out to be the best idea. And I changed this PR to use `baselayerchange`/`overlaylayerchange`

I also replaced layeradd for `map.on("layeradd",...)`, which tracks the layer's on/off state with `dataLayer.on("add", ...)`. Using `overlaylayerchange` in this case doesn't work, because this event isn't thrown when opening a new page with `...&layers=D` Plus `add ' / 'remove' looks cleaner in my opinion.

mmd-osm commented:
With the latest changes I was able to render Lake Huron locally for the first time (455k nodes). That is definitely a significant speed-up.

![image](https://github.com/user-attachments/assets/78660720-bad3-4434-a35d-7f130316edaf)

HolgerJeromin commented:
Jquery uses `JSON.parse` for parsing of `ajax()` result.

In a next step switching to `fetch" would allow using
https://developer.mozilla.org/en-US/docs/Web/API/Response/json
Where the parsing is done without blocking the main thread.

AntonKhorev commented:
We probably need to fire `baselayerchange` either from `map.setState` or from `map.updateLayers`. Otherwise if you edit `&layers=X` in the location bar, share urls are updated to its previous value.

deevroman commented:
> Otherwise if you edit &layers=X in the location bar, share urls are updated to its previous value.

Do I understand correctly that you are talking about the current behavior of the website, and not after PR? At least in Firefox, changing `&layers=X ' does not change the layers on the map, only if you open in a new tab

mmd-osm commented:
@AntonKhorev 's comment wasn't exactly clear to me either. I think it's the following. Note that once you change the layer in the URL, the share link "layer" attribute is updated accordingly:

![Peek 2025-01-14 22-58](https://github.com/user-attachments/assets/9b801345-a8e2-4a3b-ae1b-11129d9f836a)

deevroman commented:
Ah, I understand. I only tested changing the letters that represent the overlay (DNG). Changing the letters that respond to tile layers works now. PR broke it. I'll try to fix it

AntonKhorev commented:
Leaflet ships with a layers control which lets you switch between layers. That thing is responsible for [firing `baselayerchange` events](https://github.com/Leaflet/Leaflet/blob/c511e66c20dce06ac05c43cf287b3ecdba00d924/src/control/Control.Layers.js#L319). We don't use this control because we have our own layers menu. Therefore we're responsible for firing that event, which we do: https://github.com/openstreetmap/openstreetmap-website/blob/914fff9d4dec3d502d524f5b0f60797f66f8af65/app/assets/javascripts/leaflet.layers.js#L70

The problem is that our layers menu is not the only thing that can switch layers. All other places where layer change happens need also to fire `baselayerchange`. I think that `map.updateLayers` is the only such place. Lack of `baselayerchange` there becomes more noticeable with this PR because it uses this event.

Similar story with `overlaylayerchange`, although it's not a standard Leaflet event. Leaflet layers controluses `overlayadd` and `overlayremove`.

deevroman commented:
Okay, I added the `baselayerchange` firing. So, when changing D/N/G, the update does not work now, so firing `overlaylayerchange` is better done in a separate PR.

AntonKhorev commented on app/assets/javascripts/leaflet.map.js:
```diff
@@ -116,6 +116,7 @@ L.OSM.Map = L.Map.extend({
       } else {
         this.removeLayer(this.baseLayers[i]);
       }
+      this.fire("baselayerchange", { layer: this.baseLayers[i] });
```
This fires the event for every possible base layer. It's supposed to be fired only for the layer that became visible, that's how it works in `leaflet.layers.js`.

deevroman replied:
Implemented a check that there is already a layer on the map. But the code doesn’t look so attractive anymore, so it might be better to add add/remove handlers for all base layers.

AntonKhorev replied:
You could put `this.fire ...` next to `this.addLayer ...`, that's one extra line of code.

AntonKhorev commented on app/assets/javascripts/leaflet.map.js:
```diff
@@ -108,14 +108,21 @@ L.OSM.Map = L.Map.extend({
+        baselayerChanged = !this.hasLayer(this.baseLayers[i]);
         this.addLayer(this.baseLayers[i]);
       } else {
+        baselayerChanged = this.hasLayer(this.baseLayers[i]);
```
`baselayerchange` is supposed to be fired *once* for the layer that became visible. It's not supposed to be fired for removed layers. See that `null` in Leafet code?
https://github.com/Leaflet/Leaflet/blob/c511e66c20dce06ac05c43cf287b3ecdba00d924/src/control/Control.Layers.js#L317-L319

"This event would be triggered after the other layer events, and it would include a layer property set to the resulting base layer." https://github.com/Leaflet/Leaflet/pull/1064

deevroman replied:
So the next handler also violates this expectation?

https://github.com/openstreetmap/openstreetmap-website/blob/b45eb03a6c026ec0f19415100c7bb14c4c53e6cb/app/assets/javascripts/leaflet.layers.js#L62-L71

AntonKhorev replied:
The answer to the above question was *yes* but it probably didn't affect anything. However if we start using `baselayerchange` in more places, it might. For example, this code, if rewritten for `baselayerchange` wouldn't work correctly, if it expects this event to happen like Leaflet would fire it:

https://github.com/openstreetmap/openstreetmap-website/blob/b45eb03a6c026ec0f19415100c7bb14c4c53e6cb/app/assets/javascripts/index.js#L208-L216

deevroman replied:
It seems that `"baselayerchange overlaylayerchange"` will work in this place

AntonKhorev commented:
Do we fix `baselayerchange` everywhere before doing this PR that changes listeners for this event?

deevroman commented:
@mmd-osm 
> Besides, there may be some timing issue in the current code, as the behavior is not 100% reproducible for me, even today. Sometimes the map layer changes along with the share link, in other cases, the old layer is still shown.

I think I found a way to reproduce this (If you meant it):

1. Open https://www.openstreetmap.org/#map=19/60.008946/30.217791&layers=Y
2. Change Y -> H
3. Change H -> Y (the layer has not changed)


https://github.com/user-attachments/assets/31a836d2-405e-4c17-9877-c7544d6d879d

deevroman commented:
I remade PR. I abandoned `overlaylayerchange` because this event is not in Leaflet and it is not suitable in our situation.

In the case of `baselyerchange`, we can expect that our last base layer is always removed and a new base layer is added. Therefore, the removal event is not necessary to track the layer, `baslaverchange` is enough. But in the case of overlays, this is not so, overlays can just be removed from the map, without adding a new one.

However, after rejecting `layerremove`, it becomes important to maintain the invariant that there can't be multiple base layers on the map. I.e. first remove all previous layers and only then add a new one (this is important for example for a field with an iframe, sometimes you can see two base layers at the same time)

So I use `overlayadd`/`overlayremove` which are in Leaflet https://leafletjs.com/reference.html#map-event
I fire these events when the `add`/` remove` is triggered, as well as for active objects that should be shown on the map (current node, paths, note, ...)

---

What was tested:

1: Basic layer switching functionality

https://github.com/user-attachments/assets/1d7bce4e-3ab6-4366-874f-6a2bb16afb00

2. Switching layers in the URL

https://github.com/user-attachments/assets/6a201eab-65cd-44a9-804e-1efddcc3e256

3. Update links to the active object in the Edit button

https://github.com/user-attachments/assets/539fd547-09af-4d0f-8299-0b1699581a2b

AntonKhorev commented:
Merged, thanks.

mmd-osm commented:
I will create the PR to switch to JSON format shortly: https://github.com/mmd-osm/openstreetmap-website/tree/patch/datalayerjson
